src/main/java/engineering/swat/watch/impl/mac/NativeEventStream.java

```diff
@@ -0,0 +1,286 @@
+import static engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamCreateFlag.WATCH_ROOT;
+import static engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamEventFlag.ITEM_CREATED;
+import static engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamEventFlag.ITEM_INODE_META_MOD;
+import static engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamEventFlag.ITEM_MODIFIED;
+import static engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamEventFlag.ITEM_REMOVED;
+import static engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamEventFlag.MUST_SCAN_SUB_DIRS;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.Path;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import com.sun.jna.Memory;
+import com.sun.jna.Native;
+import com.sun.jna.Pointer;
+import com.sun.jna.platform.mac.CoreFoundation;
+import com.sun.jna.platform.mac.CoreFoundation.CFArrayRef;
+import com.sun.jna.platform.mac.CoreFoundation.CFIndex;
+import com.sun.jna.platform.mac.CoreFoundation.CFStringRef;
+
+import engineering.swat.watch.impl.mac.apis.DispatchObjects;
+import engineering.swat.watch.impl.mac.apis.DispatchQueue;
+import engineering.swat.watch.impl.mac.apis.FileSystemEvents;
+import engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamCallback;
+
+// Note: This file is designed to be the only place in this package where JNA is
+// used and/or the native APIs are invoked. If the need to do so arises outside
+// this file, consider extending this file to offer the required services
+// without exposing JNA and/or the native APIs.
+
+/**
+ * <p>
+ * Stream of native events for a path, issued by macOS. It's a facade-like
+ * object that hides the low-level native APIs behind a higher-level interface.
+ * </p>
+ *
+ * <p>
+ * Note: Methods {@link #open()} and {@link #close()} synchronize on this object
+ * to avoid races. The synchronization overhead is expected to be negligible, as
+ * these methods are expected to be rarely invoked.
+ * </p>
+ */
+public class NativeEventStream implements Closeable {
+
+    // Native APIs
+    private static final CoreFoundation   CF  = CoreFoundation.INSTANCE;
+    private static final DispatchObjects  DO  = DispatchObjects.INSTANCE;
+    private static final DispatchQueue    DQ  = DispatchQueue.INSTANCE;
+    private static final FileSystemEvents FSE = FileSystemEvents.INSTANCE;
+
+    // Native memory (automatically deallocated when set to `null`)
+    private @Nullable FSEventStreamCallback callback;
+    private @Nullable Pointer stream;
+    private @Nullable Pointer queue;
```

Comment:
I think they should be volatile, as their value will get ready by multiple threads, and you don't want thread-local caches. Since they could point towards bad pointers.